### PR TITLE
Reg2const arithmetic and tracking previously set ssa vars

### PR DESCRIPTION
**Pull request by RolandMacDoland**
This PR addresses the addition of two functionalities:

- `reg2const` arithmetic: this requires to keep track of registers set with a constant value through the `SetBitsOp` for potential later reuse.
- Keep track of ssa variables: in the same line as above, there is now functionality to keep track of previously created ssa vars for later reuse.

This work also encompasses an architectural re-design of the package in three main parts:

- A QIR parser
- A QIR generator
- A module that extends the `pyqir` `SimpleModule` to handle custom define quantum functions.

### CODE REVIEW

**Comment by RolandMacDoland** [comment]
Hi @cqc-alec , I recall Pete was also requesting for reviewing this PR. Since he is busy with IEEE this week, I will request his review next week.

**Review by cqc-alec — COMMENTED** [review]
Thanks. As with the other PR, I'm not sure there's much point my reviewing this until it can be tested cross-platform with public pyqir wheels.

As a general point I would like to see more tests demonstrating correctness of the conversions on (small) whole programs.

**Comment by bamarsha** [comment]
PyQIR 0.6.0 is released on PyPI now.

**Comment by RolandMacDoland** [comment]
> PyQIR 0.6.0 is released on PyPI now.

Thanks @samarsha this is going to be helpful.

**Comment by cqc-alec** [comment]
Since this includes changes from #6 I will review once that is merged.

**Review by cqc-alec — CHANGES_REQUESTED** [review]
Needs merge from `develop` please.

**Comment by RolandMacDoland** [comment]
> Needs merge from `develop` please.

Yes, sure. But since this PR contains changes from #7, I'd rather merge the other first if that's possible.

File: tests/qir_test.py
Comment (cqc-alec) [inline tests/qir_test.py]: Delete this line.

File: pytket_qir/generator.py
Comment (cqc-alec) [inline pytket_qir/generator.py]: Why this change of params?

File: pytket_qir/generator.py
Comment (cqc-alec) [inline pytket_qir/generator.py:157]: Should we handle empty registers?

File: pytket_qir/generator.py
Comment (cqc-alec) [inline pytket_qir/generator.py]: What if `in_width` is zero? Should handle edge case.

File: pytket_qir/module.py
Comment (cqc-alec) [inline pytket_qir/module.py:47]:
```suggestion
        self.gateset = gateset if gateset is not None else PYQIR_GATES
```

File: pytket_qir/module.py
Comment (cqc-alec) [inline pytket_qir/module.py]:
```suggestion
        assert self._gateset is nor None
```

File: pytket_qir/parser.py
Comment (cqc-alec) [inline pytket_qir/parser.py]: Delete.

**Review by cqc-alec — CHANGES_REQUESTED** [review]
- Please remove the 4 wheels.
- Please remove `utils.py~`.

**Comment by RolandMacDoland** [comment]
> * Please remove the 4 wheels.
> * Please remove `utils.py~`.

Thanks, I don't know why the references to these files keep propagating through various PRs although I've removed them in several other places.

File: tests/qir_test.py
Comment (RolandMacDoland) [reply tests/qir_test.py]: Thanks. Done.

File: pytket_qir/generator.py
Comment (RolandMacDoland) [reply pytket_qir/generator.py:157]: Not sure. The only calls to that method are [here](https://github.com/CQCL/pytket-qir/blob/21acfb115093a45f6dc3dcd732d070f021e9167a/pytket_qir/generator.py#L297) and [here](https://github.com/CQCL/pytket-qir/blob/21acfb115093a45f6dc3dcd732d070f021e9167a/pytket_qir/generator.py#L308) which are guaranteed to pass non-empty register.

File: pytket_qir/module.py
Comment (RolandMacDoland) [reply pytket_qir/module.py:47]: Thanks.

File: pytket_qir/module.py
Comment (RolandMacDoland) [reply pytket_qir/module.py]: Thanks

File: pytket_qir/parser.py
Comment (RolandMacDoland) [reply pytket_qir/parser.py]: Sorry, left over from debugging. Done.

File: pytket_qir/generator.py
Comment (RolandMacDoland) [reply pytket_qir/generator.py]: When I create a `SetBitsOp` with an empty register, I get a runtime error:

```python
>>> c = Circuit()
>>> b = c.add_c_register("b", 0)
>>> c.add_c_setreg(2, b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: An operation must act on at least one wire
```

This is possible though when properly creating a `SetBitsOp` 

```python
>>> c = Circuit()
>>> b = c.add_c_register("b", 3)
>>> c.add_c_setreg(2, b)
[SetBits(010) b[0], b[1], b[2]; ]
>>> com  = c.get_commands()[0]
>>> com.op
SetBits(010)
>>> com.op.n_inputs
0
>>> com.op.n_outputs
3
```

But in all cases, `n_inputs` is always 0. In this case, we are only interested in the output register which is always set otherwise we get the error above.

File: pytket_qir/gatesets/pyqir/pyqir.py
Comment (cqc-alec) [inline pytket_qir/gatesets/pyqir/pyqir.py]:
```suggestion
    base_gateset=set(_TK_TO_PYQIR.keys()),
```
Simpler, and makes it always a `set`.

File: pytket_qir/generator.py
Comment (cqc-alec) [inline pytket_qir/generator.py:114]: This may throw an exception if `auto_rebase_pass` doesn't handle the base gateset. I guess this is the best we can do, but we should probably document it somewhere.

File: tests/conftest.py
Comment (cqc-alec) [inline tests/conftest.py]: Does this need to be a fixture. as it's only used in one test?

File: pytket_qir/generator.py
Comment (cqc-alec) [reply pytket_qir/generator.py:157]: Agree about the first; not quite sure about the second: can we assume a `ClassicalExpBox` has no classical wires?

File: pytket_qir/generator.py
Comment (cqc-alec) [reply pytket_qir/generator.py]: Have you tried with latest pytket? This error was removed in https://github.com/CQCL/tket/pull/589 .

**Review by cqc-alec — CHANGES_REQUESTED** [review]

File: pytket_qir/gatesets/pyqir/pyqir.py
Comment (RolandMacDoland) [reply pytket_qir/gatesets/pyqir/pyqir.py]: Thanks, but why always ? The pipe operator won't work if the `keys()` array is empty ?

Comment (cqc-alec) [reply pytket_qir/gatesets/pyqir/pyqir.py]: > Thanks, but why always ? The pipe operator won't work if the `keys()` array is empty ?

I don't think we need the pipe operator? `keys()` does not return a `set` so with your version we get different types depending if it is empty or not.

File: pytket_qir/generator.py
Comment (RolandMacDoland) [reply pytket_qir/generator.py:114]: I am aware of the fragility of that approach. Previously, it raised an error if the gate wasn't in the translation dictionary. However, there could be scenarii (I think) where a user defines a `pytket` circuit and applies optimisations that introduce `tket` specific gates (TK1). Correct me if I am utterly wrong in that assumption though. However, one thing that you can bear in mind is that the default `pyqir` gateset is probably "standard enough" to accommodate rebases from `pytket`.  But OK I can comment on that of course, aside to letting the `NoAutoRebase` going through.

File: pytket_qir/gatesets/pyqir/pyqir.py
Comment (RolandMacDoland) [reply pytket_qir/gatesets/pyqir/pyqir.py]: Oh really ? Hm OK thanks for the clarification then.

File: pytket_qir/generator.py
Comment (RolandMacDoland) [reply pytket_qir/generator.py:114]: Done.

File: tests/conftest.py
Comment (RolandMacDoland) [reply tests/conftest.py]: Yes, that is probably a bad habit. But I try to put as much as I can in fixtures as not to clutter tests too much. Esspecially the `arrange` part.

Comment (RolandMacDoland) [reply tests/conftest.py]: Done.

File: pytket_qir/generator.py
Comment (RolandMacDoland) [reply pytket_qir/generator.py:157]: It looks impossible to add a `ClassicalExpBox` with empty registers:

```python
>>> c = Circuit(0)
>>> a = c.add_c_register("a", 0)
>>> b = c.add_c_register("b", 0)
>>> c.add_classicalexpbox_register(a ^ b, b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: An operation must act on at least one wire
```

And also: the call is within a loop that gets exectuted iff there are inputs ?

Comment (cqc-alec) [reply pytket_qir/generator.py:157]: > It looks impossible to add a `ClassicalExpBox` with empty registers:

With pytket==1.8.0rc0 it is not impossible (this was recently changed):
```
In [2]: from pytket import Circuit

In [3]: c = Circuit(0)

In [4]: a = c.add_c_register("a", 0)

In [5]: b = c.add_c_register("b", 0)

In [6]: c.add_classicalexpbox_register(a ^ b, b)
Out[6]: [ClassicalExpBox; ]
```

Comment (RolandMacDoland) [reply pytket_qir/generator.py:157]: Yes, I tried that out just before upgrading to the latest version (related to the next comment).

File: pytket_qir/generator.py
Comment (RolandMacDoland) [reply pytket_qir/generator.py]: Ok so now, it makes sense to throw an exception ? If `in_width  = 0` is valid `SetBits` that means the user is trying to assign a value to an empty register which is very likely not what it intends to do.

Comment (cqc-alec) [reply pytket_qir/generator.py]: > Ok so now, it makes sense to throw an exception ? If `in_width = 0` is valid `SetBits` that means the user is trying to assign a value to an empty register which is very likely not what it intends to do.

I don't believe we should throw an exception just because it seems like an odd thing to do. By doing so, we may force users to insert special-case handling in their own code, which will be an annoyance for them. Edge cases should be handled painlessly, if possible.

Comment (RolandMacDoland) [reply pytket_qir/generator.py]: Yes but that can be the sign of a genuine bug ? I.e. an assignment to a wrong register ?

Comment (cqc-alec) [reply pytket_qir/generator.py]: (Remember that "the user" here could be some generated code.)
Basically, if it isn't easy to handle, fine, but I'd like to understand why -- since that would probably indicate some poor design choices upstream...

Comment (RolandMacDoland) [reply pytket_qir/generator.py]: Upstream where ? In the user process ? At the `SetBits` op level ?
Furthermore, I don't see a clear way to handle that case painlessly without further understanding of the user's intentions which looks like a complicated thing to do. For instance, setting the register size to an arbitrary value if empty ? I am not too sure it makes sense. But maybe I am wrong.

Comment (cqc-alec) [reply pytket_qir/generator.py]: > Upstream where ? In the user process ? At the `SetBits` op level ? Furthermore, I don't see a clear way to handle that case painlessly without further understanding of the user's intentions which looks like a complicated thing to do. For instance, setting the register size to an arbitrary value if empty ? I am not too sure it makes sense. But maybe I am wrong.

By "upstream" I mean in some code that this depends on: that may be elsewhere in pytket-qir or it may be in pytket or it may be in pyqir. But the user can reasonably expect to be able to do this, whether or not it seems sensible. The semantics of `SetBits` with no wires are unambiguous: it is a no-op. But sometimes code generation can emit no-ops and they shouldn't cause exceptions. I'm not sure why we need to understand the user's intentions beyond this? Perhaps I am missing something.

Comment (RolandMacDoland) [reply pytket_qir/generator.py]: Thanks for this. It is clear. What I meant is how do you disambiguate between a code generation that emits a no-op as part of a process and a genuine bug ? If you could, it would justify to handle the cases appropriately: ie raising an exception or painlessly going through. If not, how do you handle the case other than a generic behaviour like raising an exception ?

Comment (cqc-alec) [reply pytket_qir/generator.py]: > Thanks for this. It is clear. What I meant is how do you disambiguate between a code generation that emits a no-op as part of a process and a genuine bug ? If you could, it would justify to handle the cases appropriately: ie raising an exception or painlessly going through. If not, how do you handle the case other than a generic behaviour like raising an exception ?

We can't, and it's not our jpb to distinguish it -- any more than it is the job of gcc to error if a line of C code has no effect. We can optimize it away if we want -- but should allow it. If it's a bug, that's someone else's problem. Am I missing the point?

Comment (RolandMacDoland) [reply pytket_qir/generator.py]: No, I think we're on the same page. But in this context, since setting a `SetBits` with empty registers is now valid, it could still be a potential bug and I don't see any other way to handle the case without raising an exception if I am to strive for correctness ?

File: pytket_qir/generator.py
Comment (RolandMacDoland) [reply pytket_qir/generator.py:157]: Done.

File: pytket_qir/generator.py
Comment (RolandMacDoland) [reply pytket_qir/generator.py]: Done.

**Review by cqc-alec — APPROVED** [review]
Could you please create a (low-priority) task to make `SetBits` and `ClassicalExpBox` support the trivial no-wires case? Ideally we should support this but maybe there are issues with pytket that prevent it.

**Comment by RolandMacDoland** [comment]
> Could you please create a (low-priority) task to make `SetBits` and `ClassicalExpBox` support the trivial no-wires case? Ideally we should support this but maybe there are issues with pytket that prevent it.

Sure. But don't they already support it ? I thought from the examples above that setting them with empty registers was valid. Or, am I getting confused ?

Thanks for the review.

**Comment by cqc-alec** [comment]
> > Could you please create a (low-priority) task to make `SetBits` and `ClassicalExpBox` support the trivial no-wires case? Ideally we should support this but maybe there are issues with pytket that prevent it.
> 
> Sure. But don't they already support it ? I thought from the examples above that setting them with empty registers was valid. Or, am I getting confused ?

E.g. https://github.com/CQCL/pytket-qir/pull/8/commits/e0c46daf5c66f8cee80f27177d1663c3bc98e8d5

I still don't get why we have to raise an exception here (I mean, I get it in terms of how the code works currently, but I don't see why it has to be that way).

Anyway this is low-priority and shouldn't block merging this!

**Comment by RolandMacDoland** [comment]
> > > Could you please create a (low-priority) task to make `SetBits` and `ClassicalExpBox` support the trivial no-wires case? Ideally we should support this but maybe there are issues with pytket that prevent it.
> > 
> > 
> > Sure. But don't they already support it ? I thought from the examples above that setting them with empty registers was valid. Or, am I getting confused ?
> 
> E.g. [e0c46da](https://github.com/CQCL/pytket-qir/commit/e0c46daf5c66f8cee80f27177d1663c3bc98e8d5)
> 
> I still don't get why we have to raise an exception here (I mean, I get it in terms of how the code works currently, but I don't see why it has to be that way).
> 
> Anyway this is low-priority and shouldn't block merging this!

Tbf, I don't understand very clearly either what you are expecting. I am a little confused between the new behaviour of the ops and handling it in the context of correctness here. Maybe best to discuss that in Slack ?

**Pull request merged**
